src/core_modules/capture-core-utils/RulesEngine/RulesEngine.js

```diff
@@ -16,80 +17,380 @@ import type {
+            } else {
+                log.warn(`Rule id:'${rule.id}'' and name:'${rule.name}' had no condition specified. Please check rule configuration.`);
+            }
+            return { isRuleEffective, rule };
+        })
+        .filter(({ isRuleEffective }) => isRuleEffective)
+        .flatMap(({ rule }) => rule.programRuleActions.map((
+            {
+                data: actionData,
+                programRuleActionType: action,
+                id,
+                location,
+                dataElementId,
+                trackedEntityAttributeId,
+                programStageId,
+                programStageSectionId,
+                optionGroupId,
+                optionId,
+                content,
+            }) => {
+            let ruleEffectData;
+            if (actionData) {
+                const strippedExpression = replaceVariablesWithValues(actionData, variablesHash);
+                const evaluatedRuleEffectData = executeExpression(dhisFunctions, strippedExpression, e => log.warn(`Expression with id rule: action:${id} could not be run. Original condition was: ${rule.condition} - Evaluation ended up as:${strippedExpression} - error message:${e}`));
+                ruleEffectData = trimQuotes(evaluatedRuleEffectData);
+            }
+            return {
+                data: ruleEffectData,
+                id,
+                location,
+                action,
+                dataElementId,
+                trackedEntityAttributeId,
+                programStageId,
+                programStageSectionId,
+                optionGroupId,
+                optionId,
+                content,
+            };
+        }));
+}
+
+
+export default class RulesEngine {
+  variableService: VariableService;
+  dateUtils: DateUtils;
+  processRulesEffects: (
+        effects: ?Array<ProgramRuleEffect>,
+        processType: $Values<typeof processTypes>,
+        dataElements: ?DataElements,
+        trackedEntityAttributes: ?TrackedEntityAttributes
+  ) => ?OutputEffects;
+  generateEffects: (
+    variablesHash: RuleVariables,
+    programRules: ?Array<ProgramRule>,
+    dataElements: ?DataElements,
+    trackedEntityAttributes: ?TrackedEntityAttributes
+  )=> ?OutputEffects;
+
+  constructor() {
+      const valueProcessor = new ValueProcessor(inputValueConverter);
+
+      this.dateUtils = getDateUtils(momentConverter);
+      this.variableService = new VariableService(valueProcessor.processValue);
+      this.processRulesEffects = getRulesEffectsProcessor(
+          convertRuleEffectDataToOutputBaseValue,
+          rulesEffectsValueConverter,
+      );
+
+      this.generateEffects = (variablesHash, programRules, dataElements, trackedEntityAttributes) => {
```

Comment:
Is there a good reason why generateEffects is defined in the constructor like this? Could we make it a static (depends on the comment above) or a prototype method?

Reply:
Yeah thanks for pointing out. I havent realised!